finality-aleph/src/network.rs

```diff
@@ -141,148 +228,195 @@ enum SessionStatus {
-    net_command_tx: mpsc::UnboundedSender<SessionCommand>,
+pub(crate) struct SessionManager<D: Clone + Encode + Decode> {
+    peer_id: PeerId,
+    sessions: Arc<Mutex<HashMap<SessionId, SessionData<D>>>>,
+    commands_for_session: mpsc::UnboundedSender<SessionCommand<D>>,
 }
 
-impl SessionManager {
-    // NOTE: later this will also need access to KeyStore :/ (for Reliable Broadcast)
+impl<D: Clone + Encode + Decode> SessionManager<D> {
     pub(crate) fn start_session(
         &self,
         session_id: SessionId,
-        _authorities: Vec<AuthorityId>,
-    ) -> RushNetwork {
-        let (net_event_tx, net_event_rx) = mpsc::unbounded();
+        keychain: KeyBox,
+    ) -> GenericNetwork<D> {
+        let (data_for_user, data_from_network) = mpsc::unbounded();
+        let auth_data = AuthData {
+            session_id,
+            peer_id: self.peer_id,
+            node_id: keychain.index(),
+        };
+        let signature = keychain.sign(&auth_data.encode());
         let session_data = SessionData {
-            net_event_tx,
+            data_for_user,
             status: SessionStatus::InProgress,
-            _authorities,
+            keychain,
+            auth_data: auth_data.clone(),
+            auth_signature: signature.clone(),
         };
         self.sessions.lock().insert(session_id, session_data);
-        RushNetwork {
+        if let Err(e) = self
+            .commands_for_session
+            .unbounded_send(SessionCommand::Meta(auth_data, signature, None))
+        {
+            log::error!(target: "afa", "sending auth command failed in new session: {}", e);
+        }
+        GenericNetwork {
             session_id,
-            net_event_rx,
-            net_command_tx: self.net_command_tx.clone(),
+            data_from_network,
+            commands_for_session: self.commands_for_session.clone(),
         }
     }
 }
 
-impl<B: BlockT + 'static, N: Network<B> + Clone> ConsensusNetwork<B, N> {
+impl<D: Clone + Encode + Decode, B: BlockT + 'static, N: Network<B> + Clone>
+    ConsensusNetwork<D, B, N>
+{
     /// Create a new instance.
-    pub(crate) fn new(network: N, protocol: impl Into<Cow<'static, str>>) -> Self {
-        let protocol = protocol.into();
-        let (net_command_tx, net_command_rx) = mpsc::unbounded();
+    pub(crate) fn new(network: N, protocol: Cow<'static, str>) -> Self {
+        let (commands_for_session, commands_from_user) = mpsc::unbounded();
         ConsensusNetwork {
             network,
             protocol,
             sessions: Arc::new(Mutex::new(HashMap::new())),
-            net_command_tx,
-            net_command_rx,
+            commands_for_session,
+            commands_from_user,
             peers: Arc::new(Mutex::new(Peers::new())),
-            phantom: PhantomData,
+            _phantom: PhantomData,
         }
     }
 
-    pub(crate) fn session_manager(&self) -> SessionManager {
+    pub(crate) fn session_manager(&self) -> SessionManager<D> {
```

Comment:
Why not store the SessionManager as a field of `ConsensusNetwork`?

Reply:
`ConsensusNetwork` is passed to a separate thread to do its work there, so `SessionManager` has to be a separate object. It might work as a `Arc<Mutex<...>>` thing, but I think at this point it would be more complicated than having a separate object.

Reply:
I mean the session manager is constructed from three fields. why not couple these 3 fields into one field of type SessionManager?

Reply:
One of them is actually taken out of `network`, so this would introduce slight data duplication, another is the channel, which not being hidden within a substruct is somewhat more readable in the code that uses it (I started writing this before noticing the thread issue). The last one, `sessions`, would actually be better within a substruct, but well, it's only one.